src/basicfuns.jl

```diff
@@ -152,9 +152,38 @@ Return `log(1+exp(x))` evaluated carefully for largish `x`.
 
 This is also called the ["softplus"](https://en.wikipedia.org/wiki/Rectifier_(neural_networks))
 transformation, being a smooth approximation to `max(0,x)`. Its inverse is [`logexpm1`](@ref).
+
+See:
+ * Martin Maechler (2012) [“Accurately Computing log(1 − exp(− |a|))”](http://cran.r-project.org/web/packages/Rmpfr/vignettes/log1mexp-note.pdf)
 """
-log1pexp(x::Real) = x < 18.0 ? log1p(exp(x)) : x < 33.3 ? x + exp(-x) : oftype(exp(-x), x)
-log1pexp(x::Float32) = x < 9.0f0 ? log1p(exp(x)) : x < 16.0f0 ? x + exp(-x) : oftype(exp(-x), x)
+function log1pexp(x::Real)
+    x0, x1, x2 = _log1pexp_thresholds(x)
+    if x ≤ x0
+        return exp(x)
+    elseif x ≤ x1
+        return log1p(exp(x))
+    elseif x ≤ x2
```

Comment:
Julia uses 1-based indexing :slightly_smiling_face: 
```suggestion
    x1, x2, x3 = _log1pexp_thresholds(x)
    if x ≤ x1
        return exp(x)
    elseif x ≤ x2
        return log1p(exp(x))
    elseif x ≤ x3
```

Reply:
BTW is it easier to check `<` instead of `<=`? So maybe use
```suggestion
    x1, x2, x3 = _log1pexp_thresholds(x)
    if x < x1
        return exp(x)
    elseif x < x2
        return log1p(exp(x))
    elseif x < x3
```

Reply:
FWIW, there is no indexing here, just notation. As long as it is documented, @cossio should just use whatever variables preferred as far as I am concerned.

Reply:
> Julia uses 1-based indexing slightly_smiling_face

Right, I was just being consistent with the notation of the paper. Since there is no name for the first bound I went with `x0`.

Reply:
Sure, I'm fine with the names, just thought it was a bit funny to point it out :smile: But if you keep them maybe change it to
```suggestion
    x0, x1, x2 = _log1pexp_thresholds(x)
    if x < x0
        return exp(x)
    elseif x < x1
        return log1p(exp(x))
    elseif x < x2
```
as it's a slightly simpler check in general (https://github.com/JuliaLang/julia/blob/5ac002038d93ea96f682bf005abc845427d5540c/base/operators.jl#L405)?

Reply:
Sure, I replaced <= with <